### PR TITLE
LibVCX V3 (Aries) Delete Credential API

### PR DESCRIPTION
**Pull request by airoasis**
Developed a new API that deletes a credential from the wallet and in the memory.
It only takes care of the Aries protocol (V3).

For now, I implemented the Java wrapper only.

### CODE REVIEW

File: vcx/libvcx/src/credential.rs [hunk 1]
```diff
@@ -530,6 +530,27 @@ pub fn get_credential(handle: u32) -> VcxResult<String> {
...
+            Credentials::V3(ref credential) => {
+                credential.delete_credential()?;
+                Ok(error::SUCCESS.code_num)
+            }
+        }
+    })
+        .map(|_| error::SUCCESS.code_num)
+        .or(Err(VcxError::from(VcxErrorKind::InvalidCredentialHandle)))
+        .and(release(handle))
+        .and_then(|_| Ok(error::SUCCESS.code_num))
```
Comment (mirgee) [inline vcx/libvcx/src/credential.rs:551]: I am not sure why the `map()` and `or()` are necessary here? How about
```
    .map_or_else(|e| Err(e), |v| {
        release(handle)?;
        Ok(v)
    })
```

File: vcx/libvcx/src/v3/handlers/issuance/holder.rs [hunk 2]
```diff
@@ -199,6 +199,16 @@ impl HolderSM {
...
         }
     }
+
+    pub fn delete_credential(&self) -> VcxResult<()> {
+        match self.state {
+            HolderState::Finished(ref state) => {
+                let cred_id = state.cred_id.clone().ok_or(VcxError::from_msg(VcxErrorKind::InvalidState, "Cannot get credential: Credential Id not found"))?;
+                _delete_credential(&cred_id)
+            }
+            _ => Err(VcxError::from_msg(VcxErrorKind::NotReady, "Cannot delete credential: Credential Issuance is not finished yet"))
```
Comment (mirgee) [inline vcx/libvcx/src/v3/handlers/issuance/holder.rs:209]: Nitpick: I don't like the capitalization, but it seems to be an issue throughout the codebase (and used inconsistently)...

Comment (mirgee) [inline vcx/libvcx/src/v3/handlers/issuance/holder.rs:209]: Maybe we should add `ErrorKind` and `ErrorCode` for these errors, not sure.

File: vcx/libvcx/src/v3/handlers/issuance/holder.rs [hunk 3]
```diff
@@ -244,6 +254,12 @@ fn _store_credential(credential: &Credential,
                                     rev_reg_def_json.as_ref().map(String::as_str))
 }
 
+fn _delete_credential(cred_id: &str) -> VcxResult<()> {
+    trace!("Holder::_delete_credential >>>");
```
Comment (mirgee) [inline vcx/libvcx/src/v3/handlers/issuance/holder.rs:258]: Please, log `cred_id`.

File: vcx/libvcx/src/credential.rs (hunk 1 above)
Comment (mirgee) [inline vcx/libvcx/src/credential.rs:551]: Why are these lines necessary? How about just
```
Credentials::V3(ref credential) => {
    credential.delete_credential()?;
    Ok(error::SUCCESS.code_num)
}
```

File: vcx/libvcx/src/api/credential.rs [hunk 4]
```diff
@@ -240,6 +240,50 @@ pub extern fn vcx_get_credential(command_handle: CommandHandle,
...
 }
 
+/// Delete a Credential from the wallet and release its handle.
+///
+/// # Params
+/// command_handle: command handle to map callback to user context.
+///
+/// credential_handle: handle of the credential to delete.
+///
+/// cb: Callback that provides feedback of the api call.
```
[the remark this answers is not in the document]
Comment (airoasis) [reply vcx/libvcx/src/api/credential.rs]: Fixed. Added "TODO: Implement" comment as well.

File: vcx/libvcx/src/api/credential.rs [hunk 5]
```diff
@@ -240,6 +240,50 @@ pub extern fn vcx_get_credential(command_handle: CommandHandle,
...
+                                        err: u32)>) -> u32 {
+    info!("vcx_delete_credential >>>");
+
+    check_useful_c_callback!(cb, VcxErrorKind::InvalidOption);
+    if !credential::is_valid_handle(credential_handle) {
+        return VcxError::from(VcxErrorKind::InvalidCredentialHandle).into()
+    }
+
+    trace!("vcx_connection_delete_connection(command_handle: {}, credential_handle: {})", command_handle, credential_handle);
+    spawn(move || {
```
[the remark this answers is not in the document]
Comment (airoasis) [reply vcx/libvcx/src/api/credential.rs]: Fixed. Good catch! Thanks.

File: vcx/libvcx/src/api/credential.rs [hunk 6]
```diff
@@ -240,6 +240,50 @@ pub extern fn vcx_get_credential(command_handle: CommandHandle,
...
+        match credential::delete_credential(credential_handle) {
+            Ok(_) => {
+                trace!("vcx_credential_delete_credential_cb(command_handle: {}, rc: {})", command_handle, error::SUCCESS.message);
+                cb(command_handle, error::SUCCESS.code_num);
+            }
+            Err(e) => {
+                trace!("vcx_credential_delete_credential_cb(command_handle: {}, rc: {})", command_handle, e);
+                cb(command_handle, e.into());
+            }
```
[the remark this answers is not in the document]
Comment (airoasis) [reply vcx/libvcx/src/api/credential.rs]: Fixed

File: vcx/libvcx/src/credential.rs [hunk 7]
```diff
@@ -530,6 +530,27 @@ pub fn get_credential(handle: u32) -> VcxResult<String> {
     })
 }
 
+pub fn delete_credential(handle: u32) -> VcxResult<u32> {
+    HANDLE_MAP.get_mut(handle, |credential| {
+        match credential {
+            Credentials::Pending(_) => {
+                Err(VcxError::from_msg(VcxErrorKind::InvalidCredentialHandle, "Cannot delete credential for Pending object"))
```
[the remark this answers is not in the document]
Comment (airoasis) [reply vcx/libvcx/src/credential.rs:541]: Fixed. Also, I have changed "get_mut" to "get"

File: vcx/libvcx/src/v3/handlers/issuance/holder.rs (hunk 2 above)
Comment (airoasis) [reply vcx/libvcx/src/v3/handlers/issuance/holder.rs]: Fixed

Comment (airoasis) [reply vcx/libvcx/src/v3/handlers/issuance/holder.rs]: I think we can reuse the existing one NotReady (Object not ready for specified action) because it is used in the get_credential (https://github.com/hyperledger/indy-sdk/blob/6ecf3dded79fde74206ec0939d2dda179784768c/vcx/libvcx/src/v3/handlers/issuance/holder.rs#L199) that has the same context.

File: vcx/libvcx/src/v3/handlers/issuance/holder.rs (hunk 3 above)
Comment (airoasis) [reply vcx/libvcx/src/v3/handlers/issuance/holder.rs]: Fixed

File: vcx/libvcx/src/credential.rs (hunk 1 above)
Comment (airoasis) [reply vcx/libvcx/src/credential.rs:559]: I couldn't release the credential object itself in the HANDLE_MAP.get code block. And I have referenced the code from delete_connection(https://github.com/hyperledger/indy-sdk/blob/6ecf3dded79fde74206ec0939d2dda179784768c/vcx/libvcx/src/connection.rs#L958)
